plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.trino.plugin.exchange.FileSystemExchangeManager.PATH_SEPARATOR;
+import static java.lang.Math.toIntExact;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static software.amazon.awssdk.core.client.config.SdkAdvancedClientOption.USER_AGENT_PREFIX;
+import static software.amazon.awssdk.core.client.config.SdkAdvancedClientOption.USER_AGENT_SUFFIX;
+
+public class S3FileSystemExchangeStorage
+        implements FileSystemExchangeStorage
+{
+    private static final String DIRECTORY_SUFFIX = "_$folder$";
+
+    private final Region region;
+    private final int multiUploadPartSize;
+    private final S3Client s3Client;
+    private final S3AsyncClient s3AsyncClient;
+
+    private final ExecutorService deleteExecutor = newCachedThreadPool(threadsNamed("s3-delete-%s"));
+
+    @Inject
+    public S3FileSystemExchangeStorage(ExchangeS3Config config)
+    {
+        if (config.getS3Region() != null) {
+            this.region = Region.of(config.getS3Region().toLowerCase(ENGLISH));
+        }
+        else {
+            this.region = null;
+        }
+        this.multiUploadPartSize = toIntExact(config.getS3UploadPartSize().toBytes());
+
+        AwsCredentialsProvider credentialsProvider = createAwsCredentialsProvider(config);
+        RetryPolicy retryPolicy = RetryPolicy.builder()
+                .numRetries(config.getS3MaxErrorRetries())
+                .build();
+        ClientOverrideConfiguration overrideConfig = ClientOverrideConfiguration.builder()
+                .retryPolicy(retryPolicy)
+                .putAdvancedOption(USER_AGENT_PREFIX, "")
+                .putAdvancedOption(USER_AGENT_SUFFIX, "Trino-exchange")
+                .build();
+
+        this.s3Client = createS3Client(credentialsProvider, overrideConfig);
+        this.s3AsyncClient = createS3AsyncClient(credentialsProvider, overrideConfig);
+    }
+
+    @Override
+    public void initialize(URI baseDirectory)
+    {
+        // Check for expiration life cycle rules
+        String bucketName = getBucketName(baseDirectory);
+        GetBucketLifecycleConfigurationRequest request = GetBucketLifecycleConfigurationRequest.builder()
+                .bucket(bucketName)
+                .build();
+        GetBucketLifecycleConfigurationResponse response = s3Client.getBucketLifecycleConfiguration(request);
+
+        verify(response.rules().stream().anyMatch(
+                rule -> rule.expiration() != null &&
+                        rule.abortIncompleteMultipartUpload() != null &&
+                        rule.status().equals(ExpirationStatus.ENABLED) &&
+                        rule.filter().equals(LifecycleRuleFilter.builder().build())
+                ), "Expected file expiration and abortIncompleteMultipartUpload lifecycle rule for exchange bucket %s", baseDirectory.toString());
```

Comment:
Not sure about it. I would not force it - you can setup GC in a different way than using AWS rules.

Reply:
I don't feel strongly about it --- but it's hard to imagine there are better ways than AWS lifecycle rules. Will leave a TODO here.